### PR TITLE
Servlet filters support

### PR DESCRIPTION
**Pull request by SAPessi**
Merging support for servlet filters into master branch. This should address issue #15. For containers that support the servlet specs you can use a `StartupHandler` to register filters in the servlet context:

```java
handler.setStartupHandler(c -> {
    FilterRegistration.Dynamic registration = c.addFilter("CustomHeaderFilter", CustomHeaderFilter.class);
    // update the registration to map to a path
    registration.addMappingForUrlPatterns(EnumSet.of(DispatcherType.REQUEST), true, "/*");
    // servlet name mappings are disabled and will throw an exception
});
```

The container does NOT load filter implementations annotated with `@WebFilter` automatically, they still have to be registered in the `onStartup` method. This is to avoid impacting cold starts.

### CODE REVIEW

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/filters/UrlPathValidator.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+
+    //-------------------------------------------------------------
+    // Implementation - Filter
+    //-------------------------------------------------------------
+
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        if (filterConfig.getInitParameter("invalid_status_code") != null) {
```
Comment (jt0) [inline aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/filters/UrlPathValidator.java]: Make "invalid_status_code" a constant.

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/filters/UrlPathValidator.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            String statusCode = filterConfig.getInitParameter("invalid_status_code");
+            try {
+                invalidStatusCode = Integer.parseInt(statusCode);
+            } catch (NumberFormatException e) {
+                invalidStatusCode = DEFAULT_ERROR_CODE;
```
Comment (jt0) [inline aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/filters/UrlPathValidator.java]: Log?

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequest.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+     * Protected constructors for implemnenting classes. This should be called first with the context received from
+     * AWS Lambda
+     * @param lambdaContext The Lambda function context. This object is used for utility methods such as log
+     */
+    AwsHttpServletRequest(Context lambdaContext) {
+        lamdaContext = lambdaContext;
+        attributes = new HashMap<>();
+
+        // TODO: We are setting this to request by default
+        dispatcherType = DispatcherType.REQUEST;
```
Comment (SAPessi) [inline aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequest.java]: Changed to static return. Cannot see where we'll use the other dispatcher types any time soon: http://docs.oracle.com/javaee/6/api/javax/servlet/DispatcherType.html

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequest.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+    /**
+     * Given the Cookie header value, parses it and creates a Cookie object
+     * @param headerValue The string value of the HTTP Cookie header
+     * @return An array of Cookie objects from the header
+     */
+    protected Cookie[] parseCookies(String headerValue) {
+        List<Cookie> output = new ArrayList<>();
+
+        for (AbstractMap.SimpleEntry<String, String> entry : this.parseHeaderValue(headerValue)) {
```
Comment (SAPessi) [inline aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequest.java]: neat trick!

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequest.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+     * is populated. For example, The header <code>Accept: application/json; application/xml</code> will contain two
+     * key value pairs with key null and the value set to application/json and application/xml respectively.
+     *
+     * @param headerContent The string value for the HTTP header
+     * @return A list of SimpleMapEntry objects with all of the possible values for the header.
+     */
+    protected List<AbstractMap.SimpleEntry<String, String>> parseHeaderValue(String headerContent) {
+        List<AbstractMap.SimpleEntry<String, String>> values = new ArrayList<>();
+        if (headerContent != null) {
+            for (String kv : headerContent.split(HEADER_VALUE_SEPARATOR)) {
```
Comment (SAPessi) [inline aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequest.java]: from a string? Will that really be faster than String.split() and loop?

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/FilterChainHolder.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    /**
+     * Get the <code>FilterHolder</code> object from the chain at the given index.
+     * @param idx The index in the chain. Use the <code>filterCount</code> method to get the filter count
+     * @return A populated FilterHolder object
+     */
+    public FilterHolder getFilter(int idx) {
+        if (filters == null) {
+            return null;
+        } else {
+            return filters.get(idx);
```
Comment (SAPessi) [inline aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/FilterChainHolder.java]: This method is not part of the specs. I think I want the exception thrown here, your code is clearly trying to do something wrong and I don't want it hidden

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/FilterChainManager.java [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+     * This method currently does not filter on servlet name because the library assumes we are using a single servlet.
+     * @param request The incoming servlet request
+     * @return A <code>FilterChainHolder</code> object that can be used to apply the filters to the request
+     */
+    public FilterChainHolder getFilterChain(final HttpServletRequest request) {
+        String targetPath = request.getServletPath();
+        DispatcherType type = request.getDispatcherType();
+
+        if (filtersSize == -1) {
+            getFilterHolders().size();
```
Comment (SAPessi) [inline aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/FilterChainManager.java]: Good catch, bug. forgot to remove.

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/filters/UrlPathValidator.java (hunk 2 above)
Comment (SAPessi) [inline aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/filters/UrlPathValidator.java]: Need to find a better log strategy. Tired of just calling the lambdaContext.getLogger from places I can't reach it. I'll add this as an issue for the next release.

**Pull request merged**